numbat/src/typechecker/mod.rs

```diff
@@ -294,17 +294,19 @@ impl TypeChecker {
             ast::Expression::Identifier(span, name) => {
                 let type_scheme = self.identifier_type(*span, name)?.clone();
 
-                let qt = type_scheme.instantiate(&mut self.name_generator);
+                let ty = match type_scheme {
+                    TypeScheme::Concrete(ty) => ty,
+                    TypeScheme::Quantified(_, _) => {
+                        let qt = type_scheme.instantiate(&mut self.name_generator);
 
-                for Bound::IsDim(t) in qt.bounds.iter() {
-                    self.constraints.add(Constraint::IsDType(t.clone())).ok();
-                }
+                        for Bound::IsDim(t) in qt.bounds.iter() {
+                            self.constraints.add(Constraint::IsDType(t.clone())).ok();
+                        }
+                        qt.inner
+                    }
+                };
```

Comment:
I’m really not sure I was allowed to do that, and I don't really understand what it does anyway, so please triple-check this part

Reply:
I'll try to explain this. Both for your understanding, but also for my own :smile:

First, we need to understand the difference between concrete and quantified types. All "normal" types like `Scalar`, `Meter`, `String`, `Bool`, `List<List<Bool>>` etc. are concrete. The distinction is only needed when type variables come into play. A generic function like
```
fn id<A>(x: A) -> A = x
```
has a quantified type (scheme) that can be written as `forall A. A -> A`. It is the same type as `forall Z. Z -> Z`. The `A` or the `Z` inside are not "concrete" types, because they are bound by the `forall` *quantifier*.

However, this does not mean that all type variables imply quantified types. During type inference, we need to *instantiate* type schemes like `forall A. A -> A` if a function (or identifier) is being used. For example, imagine someone uses the `id` function twice in an expression:
```
if id(true) then id("yes") else "no"
```
When we encounter `id(true)`, we instantiate the type scheme `forall A. A -> A` with a fresh type variable, meaning we pick a new name for a type variable that doesn't clash with anything that exists, say `T1`. The type of *this particular instance of* `id` then becomes `T1 -> T1`. A concrete type! Similarly, we instantiate the other `id` with a fresh type variable to get `T2 -> T2`; another concrete type. Constraint resolution will then find solutions for those type variables (`T1 ~ Bool`, `T2 ~ String`).

Another thing to understand is the opposite of instantiation: generalization. If you write a function like, say
```
fn f(x) = [x, x]
```
without any type annotations, the type inference algorithm will also pick a fresh type variable for `x` (say `T3`) and arrive at the concrete type `T3 -> List<T3>` for this function. As a final step *after* elaboration and constraint solving, the type checker will then perform generalization. This means it will quantify over all type variables that are not bound by the environment. So the type of `f` finally becomes `forall T3. T3 -> List<T3>`, which is equivalent to `forall A. A -> List<A>` or:
```
fn f<A>(x: A) -> List<A> = [x, x]
```
(try it [here](https://numbat.dev/?q=fn+f%28x%29+%3D+%5Bx%2C+x%5D%E2%8F%8E))

One final thing to realize is that identifiers can *also* have generic types, not just functions. For example:
```
let zero = 0
```
will lead to a type of `forall A. A` for `zero` (Side remark: there is currently no way to write down a generic type annotation for variables in Numbat).

When we use this identifier in an expression like
```
(meter + zero) / (second + zero)
```
we need to *instantiate* the type scheme `forall A. A` for each of these occurrences. In one of them, `zero` will have the type `Length`. In the other, `zero` will be of type `Time`. If we would only instantiate once and only introduce a single type variable, this code would not type check.

---

Now, back to this particular part of the code. The old code assumed that identifiers would always have quantified types. This was true before, since a variable definition would be elaborated + constraint-solved + generalized completely before anyone could use the variable in an expression.

With the change here, we can now have (multiple) variable definitions inside a single statement (the function definition). For example, if we write
```
fn f() = (meter + zero) / (second + zero)
  where zero = 0
```
the type checker will perform elaboration first. It will process the `zero = 0` variable definition and assign a fresh type variable `T4` to zero, because it will *instantiate* the type scheme for the expression `0` (which is `forall A. A`).

Next, it will try to elaborate the body of the function definition *and that will fail*. Because we did not run (constraint solving and) generalization. When `zero` is discovered in the body of the function, it will not have a quantified type of `forall A. A`, but a concrete type of `T4`.

And this is why you got *"Tried to instantiate concrete type"* errors here. The generalization procedure did not run in between the variable definition for `zero` and the elaboration of the function body.

A lot of ML-style languages solve this problem by a procedure called `let`-generalization. This complicates the type checking phase significantly. Especially with more advanced type systems like we have in Numbat (with the physical-dimension types). This is why there is actually a paper called ["Let Should not be Generalised"](https://www.microsoft.com/en-us/research/wp-content/uploads/2016/02/tldi10-vytiniotis.pdf) by no other than Simon Peyton Jones (amongst others), arguing that the above is not actually a big deal. Yes, this looks like it's too restrictive, but it turns out that no one actually writes code like that except for some very special cases.

In the original [type system extension ticket](https://github.com/sharkdp/numbat/issues/433), I already wrote down at the time that we should probably follow the suggestion of that paper, in particular since it mentions "dimension types" as one of the extensions that would be problematic for type systems with `let`-generalization.

In summary: you did everything correctly :smile:. And you were absolutely right in assuming that this was not a trivial thing and should be mentioned :+1:.

Reply:
That’s a big explanation, not gonna lie I didn't understand everything but it makes a great resource thanks for writing it down! 
I’ll probably re-read it a few times in the future 😅

Reply:
Of course. Let me know if you have any questions.